core/src/com/unciv/logic/battle/Battle.kt

```diff
@@ -42,6 +42,15 @@ object Battle {
             if(attacker.isDefeated()) return
         }
 
+        // Withdraw from melee ability
+        if (attacker is MapUnitCombatant && attacker.isMelee() && defender is MapUnitCombatant ) {
+            val withdraw = defender.unit.getUniques().firstOrNull { it.startsWith("May withdraw before melee")}
+            if (withdraw != null) {
+                if (doWithdrawFromMeleeAbility(attacker, defender, withdraw))
+                    return
+            }
+        }
+
```

Comment:
I suggest to move this check a few line below in the `else` section of the `if(defender.getUnitType().isCivilian() && attacker.isMelee()){` and ` else if (attacker.isRanged()) {` checks, before ` while (damageToDefender + damageToAttacker > 0) {`
In this way, you do not need to check `isMelee` within the function.

Reply:
Well - I opted for code clarity. isMelee() is cheap, the compiler should even be able to short-circuit the ICombatant one if it is clever and inline the whole thing? It would if this were c#. Reducing the number or cost of checks for the ability would help more. Not putting an extra transient bool into MapUnit like ignoresTerrainCost was an early decision where it was a simpler contains() and I didn't change it after adopting that variable-numbers-in-ability-keys thing I saw elsewhere. Also, this code is not too critical - more than 100 combats per turn seem unlikely.

As for the checks within the function - true, all the first three can never 'return' in the current state of this branch, they were meant in case someone uses it later from other places. I agree now, prerequisites for the function do not really need checks.

Reply:
This looks fine to me